### PR TITLE
feat(npm): migrate and deprecate unpublishSafe

### LINKED ISSUE #5265: npm changes unpublish policy that affect `unpublishSafe`
**What would you like Renovate to be able to do?**

npm will change unpublish policy.

- 24 hours -> 72 hours

and

```
In the past, we have addressed these requests on a case-by-case basis, and starting today you can now do this yourself if your package meets the following criteria:

    Has no package dependents in the npm Public Registry
    Has fewer than 300 downloads over the past week
    Has a single owner/maintainer

As before, you can still unpublish packages younger than 72 hours as long as there are no dependents in the public registry, regardless of package ownership or download count. This policy update only applies to packages older than 72 hours.
```

- [The npm Blog — Changes to npm Unpublish Policy - January 2020](https://blog.npmjs.org/post/190553543620/changes-to-npm-unpublish-policy-january-2020)
- [policies/unpublish.md at master · npm/policies](https://github.com/npm/policies/blob/master/unpublish.md)

This changes affect to renovatebot's `unpublishSafe` configuration:

- https://docs.renovatebot.com/configuration-options/#unpublishsafe

Maybe, it is releated with [prNotPendingHours](https://docs.renovatebot.com/configuration-options/#prnotpendinghours)?

**Describe the solution you'd like**

Need to change the value of [unpublishSafe](https://docs.renovatebot.com/configuration-options/#unpublishsafe) and the description.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by zharinov**
The existing npm-specific `unpublishSafe` setting will be removed and migrated instead to use `stabilityDays` (3). The `renovate/unpublish-safe` status check is also deprecated and the existing `renovate/stability-days` will be used instead.

Closes #5265

BREAKING CHANGE: The status check `renovate/unpublish-safe` will be replaced with `renovate/stability-days`. Please migrate any branch protection rules if you were relying on `renovate/unpublish-safe`.

### CODE REVIEW

File: lib/workers/branch/index.ts [hunk 1]
```diff
@@ -220,19 +220,6 @@ export async function processBranch(
...
-      config.canBeUnpublished &&
-      (config.prCreation === 'not-pending' ||
-        /* istanbul ignore next */ config.prCreation === 'status-success')
-    ) {
-      logger.debug(
-        'Skipping branch creation due to unpublishSafe + status checks'
-      );
-      return ProcessBranchResult.Pending;
-    }
-
```
Comment (zharinov) [inline lib/workers/branch/index.ts:235]: This is the part I'm not sure about

File: lib/workers/branch/status-checks.ts [hunk 2]
```diff
@@ -53,24 +53,20 @@ export async function setStability(config: StabilityConfig): Promise<void> {
...
   canBeUnpublished?: boolean;
   branchName: string;
 };
 
 export async function setUnpublishable(
   config: UnpublishableConfig
 ): Promise<void> {
-  if (!config.unpublishSafe) {
-    return;
-  }
```
Comment (zharinov) [inline lib/workers/branch/status-checks.ts:66]: Check for stabilityDays instead?

File: docs/usage/configuration-options.md [hunk 3]
```diff
@@ -1290,7 +1290,7 @@ Note that this limit is enforced on a per-repository basis.
 
 ## prNotPendingHours
 
-If you configure `prCreation=not-pending`, then Renovate will wait until tests are non-pending (all pass or at least one fails) before creating PRs. However there are cases where PRs may remain in pending state forever, e.g. absence of tests or status checks that are configure to pending indefinitely. Therefore we configure an upper limit - default 24 hours - for how long we wait until creating a PR. Note also this is the same length of time as for Renovate's own `unpublishSafe` status check for npm.
+If you configure `prCreation=not-pending`, then Renovate will wait until tests are non-pending (all pass or at least one fails) before creating PRs. However there are cases where PRs may remain in pending state forever, e.g. absence of tests or status checks that are configure to pending indefinitely. Therefore we configure an upper limit for how long we wait until creating a PR.
```
Comment (zharinov) [inline docs/usage/configuration-options.md:1293]: TODO: mention new presets for npm

File: lib/workers/branch/index.spec.ts [hunk 4]
```diff
@@ -83,12 +83,20 @@ describe('workers/branch', () => {
...
+        {
+          releaseTimestamp: new Date('2019-01-01').getTime(),
+          stabilityDays: 1,
+        },
+        {
+          releaseTimestamp: new Date().getTime(),
+          stabilityDays: 1,
+        },
+      ] as never;
+      git.branchExists.mockReturnValueOnce(false);
```
Comment (zharinov) [inline lib/workers/branch/index.spec.ts:99]: Basically, it became duplicate of test which is right below

File: lib/config/definitions.ts [hunk 5]
```diff
@@ -1062,7 +1062,7 @@ const options: RenovateOptions[] = [
     stage: 'package',
     type: 'object',
     default: {
-      unpublishSafe: false,
+      stabilityDays: 0,
```
Comment (rarkins) [inline lib/config/definitions.ts:1065]:
```suggestion
```

We don't want this overriding any default stabilityDays configured

File: lib/config/migration.ts [hunk 6]
```diff
@@ -251,6 +251,13 @@ export function migrateConfig(
               'masterIssue',
               'dependencyDashboard'
             );
+          } else if (
+            is.string(val[i]) &&
+            (val[i] === 'default:unpublishSafe' ||
```
Comment (rarkins) [inline lib/config/migration.ts:256]: I think it would most commonly be `:unpublishSafe` and not `default:unpublishSafe`. We need to migrate both patterns

File: lib/config/migration.ts [hunk 7]
```diff
@@ -251,6 +251,13 @@ export function migrateConfig(
...
+              val[i] === 'default:unpublishSafeDisabled')
```
Comment (rarkins) [inline lib/config/migration.ts:257]: Let's ignore this one instead of migrating it.

File: lib/config/migration.ts [hunk 8]
```diff
@@ -526,6 +533,26 @@ export function migrateConfig(
       delete migratedConfig.endpoints;
       isMigrated = true;
     }
+
+    if (typeof migratedConfig.unpublishSafe === 'boolean') {
```
Comment (rarkins) [inline lib/config/migration.ts:537]: Can't this be done using the key/value comparisons earlier?

File: lib/config/migration.ts [hunk 9]
```diff
@@ -526,6 +533,26 @@ export function migrateConfig(
...
+        }
+        migratedConfig.extends = ext;
+      }
+      delete migratedConfig.unpublishSafe;
+      isMigrated = true;
+    }
+
+    if (migratedConfig.extends && is.array<string>(migratedConfig.extends)) {
+      migratedConfig.extends = migratedConfig.extends.filter(
+        (val) => val !== 'npm:unpublishSafeDisabled'
```
Comment (rarkins) [inline lib/config/migration.ts:551]: I don't like this idea of adding it and then removing it

File: lib/config/presets/internal/npm.ts [hunk 10]
```diff
@@ -0,0 +1,12 @@
+import { Preset } from '../common';
+
+export const presets: Record<string, Preset> = {
+  unpublishSafe: {
+    description:
+      'Set a status check to warn when upgrades <  72 hours old might get unpublished',
```
Comment (rarkins) [inline lib/config/presets/internal/npm.ts:6]: I think we may want to remove this status check

File: lib/workers/branch/index.ts (hunk 1 above)
Comment (rarkins) [reply lib/workers/branch/index.ts:235]: Should be possible to remove this as stability days will be added instead?

**Review by rarkins — CHANGES_REQUESTED** [review]
I think we may want to remove the renovate/unpublish-safe check altogether and use only renovate/stability-days. We can make this a major release due to breaking change. @viceice @JamieMagee what do you think?

**Comment by viceice** [comment]
🤔 I've never used the `unpublishSafe` feature, so i'm ok to remove it and make a major version bump.

**Review by rarkins — CHANGES_REQUESTED** [review]
Please remove all signs of renovate/unpublish-safe as it will be complete replaced by renovate/stability-days

File: lib/config/migration.ts [hunk 11]
```diff
@@ -253,6 +253,22 @@ export function migrateConfig(
...
+            ? migratedConfig.extends.map((preset) =>
+                is.string(preset) && /^(default)?:unpublishSafe$/.test(preset)
+                  ? preset.replace(/^.*?:/, 'npm:')
+                  : preset
+              )
+            : [];
+          if (!presets.includes('npm:unpublishSafe')) {
+            presets.push('npm:unpublishSafe');
+          }
+          migratedConfig.extends = presets;
```
Comment (rarkins) [inline lib/config/migration.ts:269]:
```suggestion
          migratedConfig.extends = migratedConfig.extends || [];
          if (is.string(migratedConfig.extends)) {
            migratedConfig.extends = [migratedConfig.extends];
          }
          if (!migratedConfig.extends.includes('npm:unpublishSafe')) {
            presets.push('npm:unpublishSafe');
          }
```

File: lib/config/presets/internal/npm.ts (hunk 10 above)
Comment (rarkins) [inline lib/config/presets/internal/npm.ts:6]:
```suggestion
      'Set a status check pending for 3 days from release timestamp to guard against npm unpublishing',
```

File: lib/config/presets/internal/npm.ts [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+    npm: {
+      stabilityDays: 3,
+      prNotPendingHours: 73,
```
Comment (rarkins) [inline lib/config/presets/internal/npm.ts:9]:
```suggestion
```

File: lib/workers/branch/status-checks.ts [hunk 13]
```diff
@@ -14,10 +14,7 @@ async function setStatusCheck(
     branchName,
     context
   );
-  // Check if state needs setting
-  if (existingState === state) {
-    logger.debug(`Status check ${context} is already up-to-date`);
-  } else {
+  if (existingState !== state) {
```
Comment (rarkins) [inline lib/workers/branch/status-checks.ts:17]: Why change this?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/workers/branch/status-checks.ts (hunk 13 above)
Comment (zharinov) [reply lib/workers/branch/status-checks.ts:17]: It's tricky enough to make test for this branch, but I'll do if this one is important. Or maybe, something like this would be enough in the function first line?
```
logger.debug(`setStatusCheck(branchName = ${branchName}, state = ${state}, ...)`);

Comment (zharinov) [reply lib/workers/branch/status-checks.ts]: Never mind, it was not so hard :)

File: lib/workers/branch/status-checks.ts [hunk 14]
```diff
@@ -14,7 +14,6 @@ async function setStatusCheck(
     branchName,
     context
   );
-  // Check if state needs setting
```
Comment (zharinov) [inline lib/workers/branch/status-checks.ts:17]: But this comment is obvious, let's remove it maybe?

File: lib/config/migration.ts [hunk 15]
```diff
@@ -253,6 +253,18 @@ export function migrateConfig(
             );
           }
         }
+      } else if (key === 'unpublishSafe') {
+        isMigrated = true;
+        if (val === true) {
+          migratedConfig.extends = migratedConfig.extends || [];
+          if (is.string(migratedConfig.extends)) {
```
Comment (zharinov) [inline lib/config/migration.ts:260]: There is potential problem here, when single string requires migration but we already checked `key === 'extends' && is.array<string>(val)`. We need either coerce string to array also for previous `else if` branch, or don't coerce at all.

Comment (rarkins) [reply lib/config/migration.ts:260]: `extends` should be an array, but I think we allow/migrate a single string too. So ideally this check wouldn't be required but it's fairly easy to add it. Can you give an example of a config that wouldn't migrate correctly?

Comment (zharinov) [reply lib/config/migration.ts:260]: Given fields are processed in this order:
```json
{
  "extends": "config:application",
  "unpublishSafe": true,
}
```
It would migrate `'config:application'` to `['config:application', 'npm:unpublishSafe']`, but not to `['config:js-app', 'npm:unpublishSafe']`. But yes, it's easy to fix it.

Comment (rarkins) [reply lib/config/migration.ts:266]: OK thanks, please fix the `extends` migration in that case

File: lib/config/migration.ts [hunk 16]
```diff
@@ -237,7 +237,13 @@ export function migrateConfig(
         } else {
           migratedConfig.semanticCommitScope = null;
         }
-      } else if (key === 'extends' && is.array<string>(val)) {
```
Comment (rarkins) [inline lib/config/migration.ts:240]: Put this in a separate PR to merge first

**Review by rarkins — CHANGES_REQUESTED** [review]
Separate extends migration logic into new PR

**Review by rarkins — CHANGES_REQUESTED** [review]
Needs de-conflicting

File: docs/usage/configuration-options.md [hunk 17]
```diff
@@ -1308,7 +1308,9 @@ Note that this limit is enforced on a per-repository basis.
...
+
+Note: if the option `stabilityDays` is non-zero then Renovate will disable `prNotPendingHours` functionality.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1313]:
```suggestion
If you configure `prCreation=not-pending`, then Renovate will wait until tests are non-pending (all pass or at least one fails) before creating PRs.
However there are cases where PRs may remain in pending state forever, e.g. absence of tests or status checks that are configure to pending indefinitely.
Therefore we configure an upper limit for how long we wait until creating a PR.

Note: if the option `stabilityDays` is non-zero then Renovate will disable `prNotPendingHours` functionality.
```

This suggestion ensures we're following the one sentence per line style.

**Review by rarkins — CHANGES_REQUESTED** [review]
Now needs tests fixed

**Comment by zharinov** [comment]
Yes, I'm not sure if I need to do something with gradle buffer snapshot mismatch?

**Comment by rarkins** [comment]
@zharinov was this change intentional?

image

**Comment by zharinov** [comment]
Definitely no, I need to figure out how it was rewritten

**Comment by viceice** [comment]
Leftover from broken java test run, needs to be reverted an test should be okay again.

**Comment by zharinov** [comment]
Okay, will fix it soon

File: lib/manager/gradle-wrapper/__fixtures__/testFiles/gradle/wrapper/gradle-wrapper.properties [hunk 18]
```diff
@@ -1,5 +1,6 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-5.6.4-bin.zip
+distributionSha256Sum=038794feef1f4745c6347107b6726279d1c824f3fc634b60f86ace1e9fbd1768
+distributionUrl=https\://services.gradle.org/distributions/gradle-6.3-bin.zip
```
Comment (viceice) [inline lib/manager/gradle-wrapper/__fixtures__/testFiles/gradle/wrapper/gradle-wrapper.properties:4]:
```suggestion
distributionUrl=https\://services.gradle.org/distributions/gradle-5.6.4-bin.zip
```

**Comment by viceice** [comment]
> Okay, will fix it soon

Done 😏

**Comment by zharinov** [comment]
> > Okay, will fix it soon
> 
> Done 😏

Thanks. It just reverted back when I ran it locally, that's why I didn't push it

File: docs/usage/configuration-options.md [hunk 19]
```diff
@@ -1466,8 +1466,9 @@ Note that this limit is enforced on a per-repository basis.
 
 If you configure `prCreation=not-pending`, then Renovate will wait until tests are non-pending (all pass or at least one fails) before creating PRs.
 However there are cases where PRs may remain in pending state forever, e.g. absence of tests or status checks that are configure to pending indefinitely.
-Therefore we configure an upper limit - default 24 hours - for how long we wait until creating a PR.
-Note also this is the same length of time as for Renovate's own `unpublishSafe` status check for npm.
+Therefore we configure an upper limit for how long we wait until creating a PR.
+
+Note: if the option `stabilityDays` is non-zero then Renovate will disable `prNotPendingHours` functionality.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:1471]:
```suggestion
Note: if the option `stabilityDays` is non-zero then Renovate will disable the `prNotPendingHours` functionality.
```

Comment (zharinov) [reply docs/usage/configuration-options.md:1471]: @HonkingGoose you're the best goose I ever knew
